tools/packages/wesl/src/LowerAndEmit.ts

```diff
@@ -101,90 +119,106 @@ export function lowerAndEmitElem(e: AbstractElem, ctx: EmitContext): void {
+  );
 }
 
-export function emitRefIdent(e: RefIdentElem, ctx: EmitContext): void {
+export function emitRefIdent(e: RefIdentElem): SpannedText {
   if (e.ident.std) {
-    ctx.srcBuilder.add(e.ident.originalName, e.start, e.end);
+    return spannedText([e.start, e.end], e.ident.originalName);
   } else {
     const declIdent = findDecl(e.ident);
     const mangledName = displayName(declIdent);
-    ctx.srcBuilder.add(mangledName!, e.start, e.end);
+    return spannedText([e.start, e.end], mangledName);
   }
 }
 
-export function emitDeclIdent(e: DeclIdentElem, ctx: EmitContext): void {
+export function emitDeclIdent(e: DeclIdentElem): SpannedText {
   const mangledName = displayName(e.ident);
-  ctx.srcBuilder.add(mangledName!, e.start, e.end);
+  return spannedText([e.start, e.end], mangledName);
 }
 
-function emitAttribute(e: AttributeElem, ctx: EmitContext): void {
+function emitAttribute(e: AttributeElem, ctx: EmitContext): SpannedText {
   const { kind } = e.attribute;
-  // LATER emit more precise source map info by making use of all the spans
-  // Like the first case does
+  let span: Span = [e.start, e.end];
   if (kind === "attribute") {
     const { params } = e.attribute;
     if (params.length === 0) {
-      ctx.srcBuilder.add("@" + e.attribute.name, e.start, e.end);
+      return spannedText(span, "@", e.attribute.name);
     } else {
-      ctx.srcBuilder.add(
-        "@" + e.attribute.name + "(",
-        e.start,
-        params[0].start,
+      return spannedText(
+        span,
+        "@",
+        e.attribute.name,
+        "(",
+        ...emitParams(params, ctx),
+        ")",
       );
-      for (let i = 0; i < params.length; i++) {
-        emitContents(params[i], ctx);
-        if (i < params.length - 1) {
-          ctx.srcBuilder.add(",", params[i].end, params[i + 1].start);
-        }
-      }
-      ctx.srcBuilder.add(")", params[params.length - 1].end, e.end);
     }
   } else if (kind === "@builtin") {
-    ctx.srcBuilder.add(
-      "@builtin(" + e.attribute.param.name + ")",
-      e.start,
-      e.end,
-    );
+    return spannedText(span, "@builtin(", emitName(e.attribute.param), ")");
   } else if (kind === "@diagnostic") {
-    ctx.srcBuilder.add(
-      "@diagnostic" +
-        diagnosticControlToString(e.attribute.severity, e.attribute.rule),
-      e.start,
-      e.end,
+    return spannedText(
+      span,
+      "@diagnostic",
+      diagnosticControlToString(e.attribute.severity, e.attribute.rule),
     );
   } else if (kind === "@if") {
-    ctx.srcBuilder.add(
+    return spannedText(
+      span,
       `@if(${expressionToString(e.attribute.param.expression)})`,
-      e.start,
-      e.end,
     );
   } else if (kind === "@interpolate") {
-    ctx.srcBuilder.add(
-      `@interpolate(${e.attribute.params.map(v => v.name).join(", ")})`,
-      e.start,
-      e.end,
+    return spannedText(
+      span,
+      "@interpolate(",
+      ...emitNames(e.attribute.params),
+      ")",
```

Comment:
Look at how awesome this is! This innocent looking API actually gives us a source mapping for `@interpolate`, and for every single one of the params!

(Of course it's not perfect with regards to whitespace. That information is just not in the syntax tree.)